File: exercise/10/solution.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        if (s.empty() && p.empty()) {
            return true;
        } else if (p.empty()) {
            return false;
        }
        int pos = 0, state = 0, step = 0;
        char now;
        while (pos <= s.length()) {
            switch (state) {
                case 0:
                    if (step >= p.length()) {
                        return false;
                    }
                    now = p[step];
                    if (step + 1 == p.length() || p[step+1] != '*') {
                        state = 1;
                        step++;
                    } else {
                        state = 2;
                        step += 2;
                    }
                    break;
                case 1:
                    if (pos < s.length() && (now == s[pos] || now == '.')) {
                        pos++;
                        state = 0;
                    } else {
                        return false;
                    }
                    break;
                case 2:
                    if (pos < s.length() && (now == s[pos] || now == '.')) {
                        pos++;
                    } else {
                        state = 0;
                    }
            }
        }
        return true;
    }
};

File: exercise/10/solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        meet = [[False for i in range(n+1)] for j in range(m+1)]
        meet[m][n] = True
        for i in range(m, -1, -1):
            for j in range(n - 1, -1, -1):
                same = i < m and (s[i] == p[j] or p[j] == '.')
                if j + 1 < n and p[j+1] == '*':
                    meet[i][j] = meet[i][j+2] or same and meet[i+1][j]
                else:
                    meet[i][j] = same and meet[i+1][j+1]
        return meet[0][0]

